File: 9th_series/implement Stack using two queue/implement_stack_using_two_queue.js
let q1 = [];
let q2 = [];

// push
function push(x) {
    q1.push(x);
}

// pop
function pop() {
    let n = q1.length;

    for (let i = 0; i < n - 1; i++) {
        q2.push(q1.shift());
    }

    let ans = q1.shift();

    // swap
    let temp = q1;
    q1 = q2;
    q2 = temp;

    return ans;
}

// top
function top() {
    let n = q1.length;

    for (let i = 0; i < n - 1; i++) {
        q2.push(q1.shift());
    }

    let ans = q1.shift();
    q2.push(ans);

    // swap
    let temp = q1;
    q1 = q2;
    q2 = temp;

    return ans;
}

// empty
function empty() {
    return q1.length === 0;
}



push(10);
push(20);
push(30);
console.log(q1);
console.log("Top:", top());   // 30
console.log("after top() operation")
console.log(q1);
console.log("Pop:", pop());   // 30
console.log("after pop() operation")
console.log(q1);
console.log("Top:", top());   // 20
console.log("Empty:", empty()); // false

File: 9th_series/implement Stack using two queue/implement_stack_using_two_queue.cpp
#include <iostream>
#include <queue>
using namespace std;

queue<int> q1, q2;

// push
void push(int x) {
    q1.push(x);
}

// pop
int pop() {
    int n = q1.size();

    for (int i = 0; i < n - 1; i++) {
        q2.push(q1.front());
        q1.pop();
    }

    int ans = q1.front();
    q1.pop();

    // swap
    swap(q1, q2);

    return ans;
}

// top
int top() {
    int n = q1.size();

    for (int i = 0; i < n - 1; i++) {
        q2.push(q1.front());
        q1.pop();
    }

    int ans = q1.front();
    q1.pop();

    q2.push(ans);

    swap(q1, q2);

    return ans;
}

// empty
bool empty() {
    return q1.empty();
}
// printing
void printQueue(queue<int> q) {  // pass by value (copy)
    while (!q.empty()) {
        cout << q.front() << " ";
        q.pop();
    }
    cout << endl;
}

int main() {
    push(10);
    push(20);
    push(30);
    cout<<endl;
    printQueue(q1);
    cout << "Top: " << top() << endl;
    cout << "Pop: " << pop() << endl;
    cout << "Top: " << top() << endl;
    cout << "Empty: " << empty() << endl;

    return 0;
}